presto-main/src/main/java/com/facebook/presto/operator/aggregation/fixed_histogram/FixedDoubleBreakdownHistogram.java

```diff
@@ -0,0 +1,298 @@
+    {
+        add(value, 1.0);
+    }
+
+    public void add(double value, double weight)
+    {
+        add(value, weight, 1);
+    }
+
+    public void add(double value, double weight, long count)
+    {
+        int logicalBucketIndex = getIndexForValue(bucketCount, min, max, value);
+
+        add(logicalBucketIndex, weight, count);
+    }
+
+    private void add(int logicalBucketIndex, double weight, long count)
+    {
+        int foundIndex = lowerBoundBinarySearch(logicalBucketIndex, weight);
+        if (foundIndex < indices.length && indices[foundIndex] == logicalBucketIndex && weights[foundIndex] == weight) {
+            counts[foundIndex] += count;
+            return;
+        }
+
+        int[] newIndices = new int[indices.length + 1];
+        System.arraycopy(indices, 0, newIndices, 0, foundIndex);
+        newIndices[foundIndex] = logicalBucketIndex;
+        System.arraycopy(indices, foundIndex, newIndices, foundIndex + 1, indices.length - foundIndex);
+        indices = newIndices;
+
+        double[] newWeights = new double[weights.length + 1];
+        System.arraycopy(weights, 0, newWeights, 0, foundIndex);
+        newWeights[foundIndex] = weight;
+        System.arraycopy(weights, foundIndex, newWeights, foundIndex + 1, weights.length - foundIndex);
+        weights = newWeights;
+
+        long[] newCounts = new long[counts.length + 1];
+        System.arraycopy(counts, 0, newCounts, 0, foundIndex);
+        newCounts[foundIndex] = count;
+        System.arraycopy(counts, foundIndex, newCounts, foundIndex + 1, counts.length - foundIndex);
+        counts = newCounts;
+    }
+
+    public void mergeWith(FixedDoubleBreakdownHistogram other)
+    {
+        checkArgument(bucketCount == other.bucketCount, "bucket counts must be equal.");
+        checkArgument(min == other.min, "minimums must be equal.");
+        checkArgument(max == other.max, "maximums must be equal.");
+
+        for (int i = 0; i < other.indices.length; ++i) {
+            add(other.indices[i],
+                    other.weights[i],
+                    other.counts[i]);
+        }
+    }
+
+    @Override
+    public Iterator<BucketWeight> iterator()
+    {
+        return new Iterator<BucketWeight>() {
+            private int currentIndex;
+
+            @Override
+            public boolean hasNext()
+            {
+                return currentIndex < indices.length;
+            }
+
+            @Override
+            public BucketWeight next()
+            {
+                if (!hasNext()) {
+                    throw new NoSuchElementException();
+                }
+
+                final BucketWeight bucket = new BucketWeight(
+                        getLeftValueForIndex(bucketCount, min, max, indices[currentIndex]),
+                        getRightValueForIndex(bucketCount, min, max, indices[currentIndex]),
+                        weights[currentIndex],
+                        counts[currentIndex]);
+                ++currentIndex;
```

Comment:
`currentIndex++;`
No particular reason; it's just Presto codebase always put ++ after.